pandas/core/strings.py

```diff
@@ -502,6 +502,91 @@ def f(x):
                                dtype=object)
     return result, name
 
+def str_extractall(arr, pat, flags=0):
+    """
+    Find all non-overlapping groups in each subject
+    string in the Series using regular expression pat.
+    When each subject string of the Series has exactly
+    one match, extractall(pat) is the same as extract(pat).
+
+    Parameters
+    ----------
+    pat : string
+        Regular expression pattern
+    flags : int, default 0 (no flags)
+        re module flags, e.g. re.IGNORECASE
+
+    Returns
+    -------
+    A Series with one element for each match if there is only
+    one group in pat. Otherwise, a DataFrame with one row 
+    for each match and one column for each group.
+
+    Examples
+    --------
+    A pattern with one group will return a Series.
+    Indices with no matches will not appear in the result.
+
+    >>> S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
+    >>> S.str.extractall("[ab](?P<digit>\d)")
+       match
+    A  0        1
+       1        2
+    B  0        1
+    Name: digit, dtype: object
+    
+    A pattern with multiple groups will return a DataFrame.
+    
+    >>> S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
+            letter digit
+      match             
+    A 0          a     1
+      1          a     2
+    B 0          b     1
+    
+    Optional groups that do not match are NaN in the result.
+    
+    >>> S.str.extractall("(?P<letter>[ab])?(?P<digit>\d)")
+            letter digit
+      match             
```

Comment:
In fact the name of the index is taken from the original Series, and it could be None.

```
In [3]: S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])

In [5]: S.str.extractall("(?P<letter>[ab])?(?P<digit>\d)")
Out[5]: 
        letter digit
  match             
A 0          a     1
  1          a     2
B 0          b     1
C 0        NaN     1

In [6]: S.str.extractall("(?P<letter>[ab])?(?P<digit>\d)").index
Out[6]: 
MultiIndex(levels=[[u'A', u'B', u'C'], [0, 1]],
           labels=[[0, 0, 1, 2], [0, 1, 0, 0]],
           names=[None, u'match'])

In [7]: Sn = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])

In [12]: Sn.str.extractall("(?P<letter>[ab])?(?P<digit>\d)")
Out[12]: 
              letter digit
capital match             
A       0          a     1
        1          a     2
B       0          b     1
C       0        NaN     1

```

Do you think I should add a name to the Series used in the docstring?